File: sfm/core/estimation.py
import numpy as np

from sfm.core.geometry import skew, to_homogeneous

SAMPSON_FLOOR = 1e-30
MIN_POINTS_FUNDAMENTAL = 8
MIN_POINTS_PNP = 6
PNP_SAMPLE_SIZE = 10
CHEIRALITY_MARGIN = 0.0


def _isotropic_normalizer(x, target_scale):
    center = x.mean(axis=0)
    spread = np.sqrt(((x - center) ** 2).sum(axis=1)).mean()
    s = target_scale / spread
    dim = x.shape[1]
    t = np.eye(dim + 1)
    t[:dim, :dim] *= s
    t[:dim, dim] = -s * center
    return t


def fundamental_from_points(x1, x2):
    if len(x1) < MIN_POINTS_FUNDAMENTAL:
        raise ValueError(f"need {MIN_POINTS_FUNDAMENTAL} points, got {len(x1)}")
    t1 = _isotropic_normalizer(x1, np.sqrt(2))
    t2 = _isotropic_normalizer(x2, np.sqrt(2))
    n1 = (t1 @ to_homogeneous(x1).T).T
    n2 = (t2 @ to_homogeneous(x2).T).T
    a = np.stack([n2[:, 0] * n1[:, 0], n2[:, 0] * n1[:, 1], n2[:, 0],
                  n2[:, 1] * n1[:, 0], n2[:, 1] * n1[:, 1], n2[:, 1],
                  n1[:, 0], n1[:, 1], np.ones(len(n1))], axis=1)
    f = np.linalg.svd(a)[2][-1].reshape(3, 3)
    u, s, vt = np.linalg.svd(f)
    return t2.T @ (u @ np.diag([s[0], s[1], 0.0]) @ vt) @ t1


def sampson_distance(f, x1, x2):
    a1 = to_homogeneous(x1)
    a2 = to_homogeneous(x2)
    fx1 = (f @ a1.T).T
    ftx2 = (f.T @ a2.T).T
    num = np.einsum("ij,ij->i", a2, fx1) ** 2
    den = fx1[:, 0] ** 2 + fx1[:, 1] ** 2 + ftx2[:, 0] ** 2 + ftx2[:, 1] ** 2
    return np.sqrt(num / np.maximum(den, SAMPSON_FLOOR))


def ransac_fundamental(x1, x2, threshold, iterations, seed):
    rng = np.random.default_rng(seed)
    best_count, best_f = -1, None
    for _ in range(iterations):
        idx = rng.choice(len(x1), MIN_POINTS_FUNDAMENTAL, replace=False)
        try:
            f = fundamental_from_points(x1[idx], x2[idx])
        except np.linalg.LinAlgError:
            continue
        count = int((sampson_distance(f, x1, x2) < threshold).sum())
        if count > best_count:
            best_count, best_f = count, f
    if best_f is None:
        raise RuntimeError("ransac found no candidate")
    inliers = sampson_distance(best_f, x1, x2) < threshold
    if inliers.sum() < MIN_POINTS_FUNDAMENTAL:
        return best_f, inliers
    refined = fundamental_from_points(x1[inliers], x2[inliers])
    return refined, sampson_distance(refined, x1, x2) < threshold


def project_to_essential(e):
    u, s, vt = np.linalg.svd(e)
    mean = (s[0] + s[1]) / 2.0
    return u @ np.diag([mean, mean, 0.0]) @ vt


def triangulate_point(p1, p2, ray1, ray2):
    m = np.stack([ray1[0] * p1[2] - p1[0], ray1[1] * p1[2] - p1[1],
                  ray2[0] * p2[2] - p2[0], ray2[1] * p2[2] - p2[1]])
    x = np.linalg.svd(m)[2][-1]
    return x[:3] / x[3]


def triangulate(p1, p2, rays1, rays2):
    return np.array([triangulate_point(p1, p2, a, b) for a, b in zip(rays1, rays2)])


def _in_front(r, t, points):
    return (points[:, 2] > CHEIRALITY_MARGIN) & ((r @ points.T + t[:, None])[2] > CHEIRALITY_MARGIN)


def decompose_essential(e, rays1, rays2):
    u, _, vt = np.linalg.svd(project_to_essential(e))
    if np.linalg.det(u) < 0:
        u[:, -1] *= -1
    if np.linalg.det(vt) < 0:
        vt[-1, :] *= -1
    w = np.array([[0.0, -1.0, 0.0], [1.0, 0.0, 0.0], [0.0, 0.0, 1.0]])
    p1 = np.hstack([np.eye(3), np.zeros((3, 1))])
    best = (-1, None, None, None)
    for r in (u @ w @ vt, u @ w.T @ vt):
        for t in (u[:, 2], -u[:, 2]):
            points = triangulate(p1, np.hstack([r, t.reshape(3, 1)]), rays1, rays2)
            mask = _in_front(r, t, points)
            if mask.sum() > best[0]:
                best = (int(mask.sum()), r, t, mask)
    return best[1], best[2], best[3], best[0]


def solve_pnp(points, pixels, k):
    if len(points) < MIN_POINTS_PNP:
        raise ValueError(f"need {MIN_POINTS_PNP} points, got {len(points)}")
    u = _isotropic_normalizer(points, np.sqrt(3))
    t = _isotropic_normalizer(pixels, np.sqrt(2))
    xh = (u @ to_homogeneous(points).T).T
    uh = (t @ to_homogeneous(pixels).T).T
    rows = []
    for xi, ui in zip(xh, uh):
        zero = np.zeros(4)
        rows.append(np.concatenate([xi, zero, -ui[0] * xi]))
        rows.append(np.concatenate([zero, xi, -ui[1] * xi]))
    p = np.linalg.svd(np.array(rows))[2][-1].reshape(3, 4)
    p = np.linalg.inv(t) @ p @ u

    k_inv = np.linalg.inv(k)
    m = k_inv @ p[:, :3]
    tv = k_inv @ p[:, 3]
    if np.linalg.det(m) < 0:
        m, tv = -m, -tv
    um, sm, vtm = np.linalg.svd(m)
    r = um @ vtm
    if np.linalg.det(r) < 0:
        r, tv = -r, -tv
    return r, tv / sm.mean()


def ransac_pnp(points, pixels, k, threshold, iterations, seed):
    from sfm.core.geometry import reprojection_error

    rng = np.random.default_rng(seed)
    sample = min(max(PNP_SAMPLE_SIZE, MIN_POINTS_PNP), len(points))
    best_count, best = -1, None
    for _ in range(iterations):
        idx = rng.choice(len(points), sample, replace=False)
        try:
            r, t = solve_pnp(points[idx], pixels[idx], k)
        except (np.linalg.LinAlgError, ValueError):
            continue
        count = int((reprojection_error(k, r, t, points, pixels) < threshold).sum())
        if count > best_count:
            best_count, best = count, (r, t)
    if best is None:
        raise RuntimeError("pnp ransac found no candidate")

    inliers = reprojection_error(k, best[0], best[1], points, pixels) < threshold
    if inliers.sum() < MIN_POINTS_PNP:
        return best[0], best[1], inliers
    try:
        r, t = solve_pnp(points[inliers], pixels[inliers], k)
    except (np.linalg.LinAlgError, ValueError):
        return best[0], best[1], inliers
    refined = reprojection_error(k, r, t, points, pixels) < threshold
    if refined.sum() < inliers.sum():
        return best[0], best[1], inliers
    return r, t, refined

File: web/src/estimation.js
import { Matrix, SingularValueDecomposition, determinant, inverse } from '../vendor/ml-matrix.js';

import { apply, identity, matmul, normalizePixel, skew, transpose } from './geometry.js';

export const SAMPSON_FLOOR = 1e-30;
export const MIN_POINTS_FUNDAMENTAL = 8;
export const MIN_POINTS_PNP = 6;
export const PNP_SAMPLE_SIZE = 10;

const nullVector = (rows) => {
  const width = rows[0].length;
  const padded = rows.length >= width
    ? rows
    : [...rows, ...Array.from({ length: width - rows.length }, () => new Array(width).fill(0))];
  const svd = new SingularValueDecomposition(new Matrix(padded));
  return svd.rightSingularVectors.getColumn(width - 1);
};

function isotropicNormalizer(points, targetScale) {
  const dim = points[0].length;
  const center = points[0].map((_, j) => points.reduce((sum, p) => sum + p[j], 0) / points.length);
  const spread = points.reduce((sum, p) =>
    sum + Math.hypot(...p.map((v, j) => v - center[j])), 0) / points.length;
  const s = targetScale / spread;
  const t = Array.from({ length: dim + 1 }, (_, i) =>
    Array.from({ length: dim + 1 }, (_, j) => (i === j ? 1 : 0)));
  for (let i = 0; i < dim; i += 1) {
    t[i][i] = s;
    t[i][dim] = -s * center[i];
  }
  return t;
}

const homogeneous = (p) => [...p, 1];

export function fundamentalFromPoints(left, right) {
  if (left.length < MIN_POINTS_FUNDAMENTAL) {
    throw new Error(`need ${MIN_POINTS_FUNDAMENTAL} points, got ${left.length}`);
  }
  const t1 = isotropicNormalizer(left, Math.SQRT2);
  const t2 = isotropicNormalizer(right, Math.SQRT2);
  const rows = left.map((point, i) => {
    const a = apply(t1, homogeneous(point));
    const b = apply(t2, homogeneous(right[i]));
    return [b[0] * a[0], b[0] * a[1], b[0], b[1] * a[0], b[1] * a[1], b[1], a[0], a[1], 1];
  });
  const f = nullVector(rows);
  const svd = new SingularValueDecomposition(new Matrix([f.slice(0, 3), f.slice(3, 6), f.slice(6, 9)]));
  const rank2 = svd.leftSingularVectors
    .mmul(Matrix.diag([svd.diagonal[0], svd.diagonal[1], 0]))
    .mmul(svd.rightSingularVectors.transpose())
    .to2DArray();
  return matmul(matmul(transpose(t2), rank2), t1);
}

export function sampsonDistance(f, left, right) {
  return left.map((point, i) => {
    const a = homogeneous(point);
    const b = homogeneous(right[i]);
    const fa = apply(f, a);
    const ftb = apply(transpose(f), b);
    const numerator = (b[0] * fa[0] + b[1] * fa[1] + b[2] * fa[2]) ** 2;
    const denominator = fa[0] ** 2 + fa[1] ** 2 + ftb[0] ** 2 + ftb[1] ** 2;
    return Math.sqrt(numerator / Math.max(denominator, SAMPSON_FLOOR));
  });
}

function makeRandom(seed) {
  let state = seed >>> 0;
  return () => {
    state = (state * 1664525 + 1013904223) >>> 0;
    return state / 4294967296;
  };
}

function sampleIndices(count, size, random) {
  const pool = Array.from({ length: count }, (_, i) => i);
  for (let i = 0; i < size; i += 1) {
    const j = i + Math.floor(random() * (count - i));
    [pool[i], pool[j]] = [pool[j], pool[i]];
  }
  return pool.slice(0, size);
}

export function ransacFundamental(left, right, threshold, iterations, seed) {
  const random = makeRandom(seed);
  let bestCount = -1;
  let bestF = null;
  for (let i = 0; i < iterations; i += 1) {
    const idx = sampleIndices(left.length, MIN_POINTS_FUNDAMENTAL, random);
    let candidate;
    try {
      candidate = fundamentalFromPoints(idx.map((n) => left[n]), idx.map((n) => right[n]));
    } catch {
      continue;
    }
    const count = sampsonDistance(candidate, left, right).filter((d) => d < threshold).length;
    if (count > bestCount) {
      bestCount = count;
      bestF = candidate;
    }
  }
  if (!bestF) throw new Error('ransac found no candidate');
  let inliers = sampsonDistance(bestF, left, right).map((d) => d < threshold);
  if (inliers.filter(Boolean).length < MIN_POINTS_FUNDAMENTAL) return { f: bestF, inliers };
  const refined = fundamentalFromPoints(
    left.filter((_, i) => inliers[i]),
    right.filter((_, i) => inliers[i]),
  );
  inliers = sampsonDistance(refined, left, right).map((d) => d < threshold);
  return { f: refined, inliers };
}

export function projectToEssential(e) {
  const svd = new SingularValueDecomposition(new Matrix(e));
  const mean = (svd.diagonal[0] + svd.diagonal[1]) / 2;
  return svd.leftSingularVectors
    .mmul(Matrix.diag([mean, mean, 0]))
    .mmul(svd.rightSingularVectors.transpose())
    .to2DArray();
}

export function triangulatePoint(p1, p2, ray1, ray2) {
  const rows = [
    p1[2].map((v, i) => ray1[0] * v - p1[0][i]),
    p1[2].map((v, i) => ray1[1] * v - p1[1][i]),
    p2[2].map((v, i) => ray2[0] * v - p2[0][i]),
    p2[2].map((v, i) => ray2[1] * v - p2[1][i]),
  ];
  const x = nullVector(rows);
  return [x[0] / x[3], x[1] / x[3], x[2] / x[3]];
}

const asProjection = (r, t) => r.map((row, i) => [...row, t[i]]);

export function decomposeEssential(e, rays1, rays2) {
  const svd = new SingularValueDecomposition(new Matrix(projectToEssential(e)));
  const u = svd.leftSingularVectors;
  const v = svd.rightSingularVectors;
  if (determinant(u) < 0) u.mulColumn(2, -1);
  if (determinant(v) < 0) v.mulColumn(2, -1);
  const uArray = u.to2DArray();
  const vt = v.transpose().to2DArray();
  const w = [[0, -1, 0], [1, 0, 0], [0, 0, 1]];
  const translation = uArray.map((row) => row[2]);
  const p1 = asProjection(identity(), [0, 0, 0]);

  let best = { count: -1 };
  for (const r of [matmul(matmul(uArray, w), vt), matmul(matmul(uArray, transpose(w)), vt)]) {
    for (const sign of [1, -1]) {
      const t = translation.map((value) => value * sign);
      const p2 = asProjection(r, t);
      const mask = rays1.map((ray, i) => {
        const point = triangulatePoint(p1, p2, ray, rays2[i]);
        const inSecond = apply(r, point).map((value, j) => value + t[j]);
        return point[2] > 0 && inSecond[2] > 0;
      });
      const count = mask.filter(Boolean).length;
      if (count > best.count) best = { count, r, t, mask };
    }
  }
  return best;
}

export function solvePnp(points, pixels, k, kInverse) {
  if (points.length < MIN_POINTS_PNP) {
    throw new Error(`need ${MIN_POINTS_PNP} points, got ${points.length}`);
  }
  const u = isotropicNormalizer(points, Math.sqrt(3));
  const t = isotropicNormalizer(pixels, Math.SQRT2);
  const rows = [];
  points.forEach((point, i) => {
    const x = apply(u, homogeneous(point));
    const uv = apply(t, homogeneous(pixels[i]));
    rows.push([...x, 0, 0, 0, 0, ...x.map((v) => -uv[0] * v)]);
    rows.push([0, 0, 0, 0, ...x, ...x.map((v) => -uv[1] * v)]);
  });
  const solution = nullVector(rows);
  const normalized = new Matrix([solution.slice(0, 4), solution.slice(4, 8), solution.slice(8, 12)]);
  const p = inverse(new Matrix(t)).mmul(normalized).mmul(new Matrix(u)).to2DArray();

  let m = matmul(kInverse, p.map((row) => row.slice(0, 3)));
  let tv = apply(kInverse, p.map((row) => row[3]));
  if (determinant(new Matrix(m)) < 0) {
    m = m.map((row) => row.map((v) => -v));
    tv = tv.map((v) => -v);
  }
  const svd = new SingularValueDecomposition(new Matrix(m));
  let r = svd.leftSingularVectors.mmul(svd.rightSingularVectors.transpose()).to2DArray();
  if (determinant(new Matrix(r)) < 0) {
    r = r.map((row) => row.map((v) => -v));
    tv = tv.map((v) => -v);
  }
  const scale = svd.diagonal.reduce((sum, v) => sum + v, 0) / svd.diagonal.length;
  return { r, t: tv.map((v) => v / scale) };
}